tests/functional/test_new_resolver.py

```diff
@@ -2272,6 +2272,90 @@ def test_new_resolver_dont_backtrack_on_extra_if_base_constrained(
+@pytest.mark.parametrize("two_extras", (True, False))
+def test_new_resolver_dont_backtrack_on_extra_if_base_constrained_in_requirement(
+    script: PipTestEnvironment, swap_order: bool, two_extras: bool
+) -> None:
+    """
+    Verify that a requirement with a constraint on a package (either on the base
+    on the base with an extra) causes the resolver to infer the same constraint for
+    any (other) extras with the same base.
+
+    :param swap_order: swap the order the install specifiers appear in
+    :param two_extras: also add an extra for the constrained specifier
+    """
+    create_basic_wheel_for_package(script, "dep", "1.0")
+    create_basic_wheel_for_package(
+        script, "pkg", "1.0", extras={"ext1": ["dep"], "ext2": ["dep"]}
+    )
+    create_basic_wheel_for_package(
+        script, "pkg", "2.0", extras={"ext1": ["dep"], "ext2": ["dep"]}
+    )
+
+    to_install: tuple[str, str] = (
+        "pkg[ext1]",
+        "pkg[ext2]==1.0" if two_extras else "pkg==1.0",
+    )
+
+    result = script.pip(
+        "install",
+        "--no-cache-dir",
+        "--no-index",
+        "--find-links",
+        script.scratch_path,
+        *(to_install if not swap_order else reversed(to_install)),
+    )
+    assert "pkg-2.0" not in result.stdout, "Should not try 2.0 due to constraint"
+    script.assert_installed(pkg="1.0", dep="1.0")
+
+
+@pytest.mark.parametrize("swap_order", (True, False))
+@pytest.mark.parametrize("two_extras", (True, False))
+def test_new_resolver_dont_backtrack_on_conflicting_constraints_on_extras(
+    script: PipTestEnvironment, swap_order: bool, two_extras: bool
+) -> None:
+    """
+    Verify that conflicting constraints on the same package with different
+    extras cause the resolver to trivially reject the request rather than
+    trying any candidates.
+
+    :param swap_order: swap the order the install specifiers appear in
+    :param two_extras: also add an extra for the second specifier
+    """
+    create_basic_wheel_for_package(script, "dep", "1.0")
+    create_basic_wheel_for_package(
+        script, "pkg", "1.0", extras={"ext1": ["dep"], "ext2": ["dep"]}
+    )
+    create_basic_wheel_for_package(
+        script, "pkg", "2.0", extras={"ext1": ["dep"], "ext2": ["dep"]}
+    )
+
+    to_install: tuple[str, str] = (
+        "pkg[ext1]>1",
+        "pkg[ext2]==1.0" if two_extras else "pkg==1.0",
+    )
+
+    result = script.pip(
+        "install",
+        "--no-cache-dir",
+        "--no-index",
+        "--find-links",
+        script.scratch_path,
+        *(to_install if not swap_order else reversed(to_install)),
+        expect_error=True,
+    )
+    assert (
+        "pkg-2.0" not in result.stdout or "pkg-1.0" not in result.stdout
+    ), "Should only try one of 1.0, 2.0 depending on order"
+    assert (
+        "looking at multiple versions" not in result.stdout
+    ), "Should not have to look at multiple versions to conclude conflict"
```

Comment:
You make a good point. Alas, given that this runs as a subprocess, I don't see a 100% reliable approach to test this, short of hacking the subprocess' entrypoint to override the reporter passed to resolvelib.

I see two alternatives that, while imperfect, I think should be sufficiently stable:
1. rely on the well-structured output of `PipDebuggingReporter` instead of the default `PipReporter`. I would expect that one to remain stable.
2. Add a coupled (refer to this one in its docstring test case / parametrization that does look back across versions and asserts the presence of the messages. If the format ever changes, that test will fail and this one can be updated.

I personally have a big preference for alternative 1 over the other two approaches so I'll implement that now but I can easily change it if you have another suggestion/preference.